### PR TITLE
Project: Modal workflow selection

### PR DESCRIPTION
**Pull request by eatyourgreens**
A continuation of #1966. This rewrites the workflow selector to allow us to have different behaviour on the project home page or on the project classify page. On the classify page, we'd like workflow selection to interrupt you with a modal dialog that forces you to choose a workflow, and optional subject set, before continuing to classify.

[InVision prototype](https://projects.invisionapp.com/d/main#/console/12924056/440898093/preview).

Entering a project via the general Classify URL should prompt to choose a workflow:
http://localhost:3000/projects/msalmon/hms-nhs-the-nautical-health-service/classify?env=production
Screenshot of the Classify page for HMS NHS, showing a choice of workflows in a modal popup.

Entering a project via the URL of a workflow that requires a subject set should prompt for subject set selection:
http://localhost:3000/projects/msalmon/hms-nhs-the-nautical-health-service/classify/workflow/16807?env=production
Screenshot of the Classify page for HMS NHS, showing a choice of subject sets in a modal popup.

Neither of those pages should open the workflow tutorial in a popup.

Package:
app-project

# Review Checklist

## General

- [x] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [x] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [x] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [x] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

### CODE REVIEW

**Comment by eatyourgreens** [comment]
Still to do: new tests and stories. Maybe abstract some of the repeated code into new components.

**Comment by eatyourgreens** [comment]
In local testing with the HMS NHS project, I’m seeing an intermittent bug where the classifier opens a tutorial over the subject set picker, on page load. There might need to be some changes to the classifier to make sure it isn’t trying to open modals when there is already a modal open on the Classify page. Having two popups open is going to be particularly confusing in a screen reader.
Screenshot of the classifier showing a modal tutorial open at the same time as workflow selection modal.

**Comment by srallen** [comment]
The tutorial load observes and reacts to the workflow loading, so I think its store could be modified to instead observe and react to the subject load?

**Comment by eatyourgreens** [comment]
Thanks! I think that would definitely fix this. The Classify page could also hold off on passing initial props to the classifier until `canClassify` is true.

**Review by beckyrother — APPROVED** [review]
I think this works! Grant mentioned it in Slack but reiterating here that the close buttons do still need to be removed. Everything else looks great!

**Comment by eatyourgreens** [comment]
Thanks for reminding me about the Close buttons.

**Review by shaunanoordin — CHANGES_REQUESTED** [review]
## PR Review

Package: `lib-classifier`

This PR update shows the Workflow Selection modal on the `/classify` path, when a project has multiple active Workflows. Also, this update shows the Subject Selection modal on the `/classify/workflow/1234` path, when a Workflow has multiple active Subjects.

Unfortunately, I am encountering major issues while testing this PR - @eatyourgreens was this branch recently rebased to reflect changes made in #1963 ? I'm having issues while testing this.

### Major Issue: "No Workflow ID" error"

Issue: whenever I try to go to the `/classify` page of a Project with multiple active workflows (note: no default workflow), I will see the modal for 2 seconds, before the website crashes with the error `ReferenceError: No workflow ID available`

Testing setup:
- Any project with multiple active workflows, on localhost
  - e.g.`http://localhost:3000/projects/msalmon/hms-nhs-the-nautical-health-service/classify?env=production` with 2 active workflows
  - note: I used `http://localhost:3000/projects/darkeshard/transformers/classify` with 2 active workflows. but please note that this project setup will be changing rapidly as I review different PRs, so best test on your own projects.
- macOS 10 + Chrome 87

Testing steps:
- go to the `/classify` route of the selected test project
- Expected: the Workflow Selection modal should show, and the Classifier should halt all operations until I select my workflow
- Actual: the Workflow Selection modal shows for a few seconds. Then, the app crashes with the on-screen error - "Unhandled Runtime Error - Reference Error: No Workflow ID Available"
  Screen Shot 2021-01-19 at 23 12 03

Analysis:
- From a brief debug of the code, it seems that while the classifier correctly identifies when to serve the Workflow Selection modal, the "try to fetch a workflow" logic is STILL going on, and it's trying to fetch Workflow "undefined".
- The "trying to fetch workflow undefined" action might explain why there's a few second delay between seeing modal and the app crashing.

### Status

Issues detected, so I'm marking this review as a CR for the time being...

...BUT after digging into this further, I realise that the issue might be _separate from this PR._ After all, this PR adds a new Workflow Selection component (which seems to work great, even though my full testing capability is limited due to the issue) while the issue seems to stem from a _workflow selection logic._

Hence, the question: _was this PR rebased after 1963?_ I was looking at [my tests](https://github.com/zooniverse/front-end-monorepo/pull/1963#pullrequestreview-571536450) from 1963 - specifically Scenario 4 - and I realise that the "No Workflow ID" error was encountered there.

> Scenario 4: project has multiple active workflows, none of which is the default
> - the "Classification" link in the header leads to: /projects/darkeshard/transformers/classify (This is sensible)
> - navigating directly to the /classify route results in: ERROR, No Workflow ID available (Which IS expected 👍 )

So, thoughts: I think this PR (i.e. add new component) works fine in theory (code looks good), but I can't approve it until I've fully tested it, and that requires a fix to the workflow selection logic, which should bea separate PR.

**Comment by shaunanoordin** [comment]
Footnote: I'm now reviewing 1966 after 1976. I'm not sure if getting 1966 merged first will address the "No Workflow ID" error described in the PR Review, since - at the time of writing - 1966 has conflicts that need resolving.

**Comment by eatyourgreens** [comment]
@shaunanoordin #1966 is the 'quick fix' to get workflow selection onto the Classify page so that we know we can use it in the Engaging Corwds beta. This PR is the more considered approach, using the final design and an interstitial popup, but I kept it seperate just in case it turned out that the classifier breaks when you render it before workflow selection is complete. I think definitely review #1966 first, then have a look at this PR.

So #1966 is essential for the beta test. This is nice-to-have, if we can get it working in time.

**Comment by eatyourgreens** [comment]
Some [`git rebase --onto` magic](https://blog.frankel.ch/multiple-usages-git-rebase-onto/) should fix those merge conflicts.

**Comment by eatyourgreens** [comment]
Re. the workflow ID error, we have a design for a placeholder classifier that should show while we’re choosing a workflow, subject set etc. I’ll open an issue for that.

**Comment by eatyourgreens** [comment]
That error you're seeing is being thrown by `selectWorkflow` here, because we've mounted the classifier without a workflow, and we don't select a default workflow for you now. You have to make a choice from the menu.
https://github.com/zooniverse/front-end-monorepo/blob/9300aad275e211918f2f09b00eeb4a13e9c21343/packages/lib-classifier/src/components/Classifier/Classifier.js#L68

**Comment by eatyourgreens** [comment]
@shaunanoordin I've fixed the workflow ID errors. This PR is now a must-have for the Engaging Crowds beta test.

File: packages/app-project/src/screens/ClassifyPage/ClassifyPage.js [hunk 1]
```diff
@@ -37,16 +56,36 @@ function ClassifyPage (props) {
       >
 
         <Box as='main' fill='horizontal'>
+          {!canClassify && (
+            <Modal
+              active
+              closeFn={onClose}
+              onEsc={onClose}
```
Comment (mcbouslog) [inline packages/app-project/src/screens/ClassifyPage/ClassifyPage.js:63]: I think `closeFn` covers on escape, this might be able to be deleted

**Review by mcbouslog — APPROVED** [review]
Code LGTM, no blocking comments. Nice separation of components, changes to `Classifier.js` straightforward. Tests and stories are very helpful.

I see some SubjectSetCard propType warnings (expected ___ received `undefined`).

Confirming as already noted that `/classify/workflow` initial load had tutorial modal on top of subject selection modal. Not sure if there's a related issue/PR, but not blocking.

In addition to production project noted in comments, tested from project home and `/classify` on staging projects `projects/nora-dot-test/planet-finders-beta` and  `projects/wgranger-test/anti-slavery-testing`, without issues/warnings specific to this PR.

File: packages/app-project/src/screens/ClassifyPage/ClassifyPage.js (hunk 1 above)
Comment (eatyourgreens) [reply packages/app-project/src/screens/ClassifyPage/ClassifyPage.js:63]: I think you're right.

File: packages/app-project/src/screens/ProjectHomePage/components/Hero/components/WorkflowMenu/WorkflowMenu.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    <WorkflowSelector
+      onSelect={onSelectWorkflow}
+      workflows={workflows}
+    />
+    {activeWorkflow &&
+      <Modal
+        active
+        closeFn={onClose}
+        headingBackground='brand'
+        title={activeWorkflow.displayName || 'Choose a subject set'}
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/app-project/src/screens/ProjectHomePage/components/Hero/components/WorkflowMenu/WorkflowMenu.js:38]: Good catch!

**Comment by eatyourgreens** [comment]
Thanks for the review @mcbouslog. I'll have a look at fixing the tutorial on this branch but it might need its own PR.

File: packages/app-project/src/shared/components/SubjectSetPicker/SubjectSetPicker.js [hunk 3]
```diff
@@ -21,8 +21,20 @@ const StyledHeading = styled(Heading)`
   line-height: 100%;
 `
 
+function BackButton({ onClick }) {
+  return (
+    <Button
+      plain
+      onClick={onClick}
+    >
+      <SpacedText color='accent-2'>
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/app-project/src/shared/components/SubjectSetPicker/SubjectSetPicker.js:30]: I think there might be a ZRC component I can use here.
https://zooniverse.github.io/front-end-monorepo/@zooniverse/react-components/index.html?path=/story/plainbutton--light-theme-default

**Pull request merged**